File: apps/token-app-api/database/models/Order.js
const baseSchema = require("../schemas/baseSchema");
const { v4: uuid } = require('uuid');

module.exports = (sequelize, DataTypes) => {
  const Order = sequelize.define("Order",
    {
      ...baseSchema,
      status: {
        field: 'status',
        type: DataTypes.ENUM([0, 1, 2, 3, 4]),
        defaultValue: 0,
        comment: '訂單狀態: 0: 等待匯款, 1: 已匯款未放行, 2: 已放行, 3: 買家已取消, 4: 賣家已取消, 5: 後台取消',
        get() {
          const rawValue = this.getDataValue('status');
          return parseInt(rawValue);
        },
      },
      amount: {
        filed: 'amount',
        type: DataTypes.FLOAT,
        allowNull: false,
        comment: '進行交易的數量',
        get() {
          const rawValue = this.getDataValue('amount');
          return Number(rawValue);
        },
      },
      cancelReason: {
        filed: 'cancel_reason',
        type: DataTypes.STRING,
        allowNull: true,
        comment: '取消的原因',
      },
      expectedFinishAt: {
        filed: 'expected_finish_at',
        type: DataTypes.DATE,
        allowNull: false,
        comment: '預計完成交易的時間',
        get() {
          const rawValue = this.getDataValue('finishAt');
          return rawValue ? rawValue.getTime() : null;
        }
      },
      finishAt: {
        filed: 'finish_at',
        type: DataTypes.DATE,
        allowNull: true,
        comment: '完成交易的時間',
        get() {
          const rawValue = this.getDataValue('finishAt');
          return rawValue ? rawValue.getTime() : null;
        }
      },
    }, {
    sequelize,
    tableName: "orders",
    timestamps: false,
    underscored: true,
    freezeTableName: true,
    paranoid: true,
  });

  Order.beforeCreate(order => {
    const orderId = uuid();
    order.id = orderId.replace('-', '');
  });

  Order.associate = function (models) {
    // associations can be defined here
    Order.belongsTo(models.User, {
      as: "user",
      foreignKey: {
        name: 'user_id',
      },
      onUpdate: 'CASCADE',
      onDelete: 'CASCADE',
    });

    Order.belongsTo(models.PendingOrder, {
      as: "pendingOrder",
      foreignKey: {
        name: 'pending_order_id',
      },
      onUpdate: 'CASCADE',
      onDelete: 'CASCADE',
    });

    Order.belongsTo(models.Bankcard, {
      as: "bankcard",
      foreignKey: {
        name: 'bank_card_id',
      },
      onUpdate: 'CASCADE',
      onDelete: 'CASCADE',
    });
  };
  return Order;
};

File: pkg/models/order.go
package models

import (
	"time"

	"github.com/google/uuid"
)

/**
 * @brief Order 訂單資料模型
 * @description 對應資料庫 orders 表
 */
type Order struct {
	ID               uuid.UUID  `gorm:"column:id;type:uuid;primaryKey" json:"id"`
	UserID           int        `gorm:"column:user_id;not null" json:"user_id"`
	PendingOrderID   uuid.UUID  `gorm:"column:pending_order_id;type:uuid;not null" json:"pending_order_id"`
	BankCardID       int        `gorm:"column:bank_card_id;not null" json:"bank_card_id"`
	Status           int        `gorm:"column:status;not null" json:"status"`
	Amount           float64    `gorm:"column:amount;not null" json:"amount"`
	CancelReason     *string    `gorm:"column:cancel_reason;type:varchar(255)" json:"cancel_reason,omitempty"`
	ExpectedFinishAt time.Time  `gorm:"column:expected_finish_at;type:timestamptz;not null" json:"expected_finish_at"`
	FinishAt         *time.Time `gorm:"column:finish_at;type:timestamptz" json:"finish_at,omitempty"`
	CreatedAt        time.Time  `gorm:"column:created_at;type:timestamptz;not null" json:"created_at"`
	UpdatedAt        time.Time  `gorm:"column:updated_at;type:timestamptz;not null" json:"updated_at"`
	DeletedAt        *time.Time `gorm:"column:deleted_at;type:timestamptz" json:"deleted_at,omitempty"`

	// 關聯
	User         *User         `gorm:"foreignKey:UserID;references:ID" json:"user,omitempty"`
	PendingOrder *PendingOrder `gorm:"foreignKey:PendingOrderID;references:ID" json:"pending_order,omitempty"`
	BankCard     *BankCard     `gorm:"foreignKey:BankCardID;references:ID" json:"bank_card,omitempty"`
}

/**
 * @brief TableName 指定資料表名稱
 * @return string 資料表名稱
 */
func (Order) TableName() string {
	return "orders"
}
